File: app/src/main/java/com/blog/demo/custom/widget/CListView.java
package com.blog.demo.custom.widget;

import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.widget.ListView;
import android.widget.AbsListView;

import com.blog.demo.LogTool;

public class CListView extends ListView implements AbsListView.OnScrollListener {
    private CListViewHeaderView mHeaderView;
    private int mFirstItemIndex;

    private float mStartY, mCurrentY;
    private boolean mRefreshEnable, mRecord;

    public CListView(Context context) {
        this(context, null);
    }

    public CListView(Context context, AttributeSet attrs) {
        super(context, attrs);

        mHeaderView = new CListViewHeaderView(context);
        mHeaderView.setListView(this);
        addHeaderView(mHeaderView);
        mHeaderView.refreshFinish();

        setOnScrollListener(this);
    }

    @Override
    public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCount, int totalItemCount) {
        mFirstItemIndex = firstVisibleItem;
    }

    @Override
    public void onScrollStateChanged(AbsListView view, int scrollState) {
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                // 保证触摸的y值只记录一次
                if (!mRecord) {
                    mStartY = event.getY();
                    mRecord = true;
                }
                break;
            case MotionEvent.ACTION_MOVE:
                if (!mRecord) {
                    mStartY = event.getY();
                    mRecord = true;
                }
                mCurrentY = event.getY();
                LogTool.logi("CListView", "MOVE startY = " + mStartY + ", currentY = " + mCurrentY);
                if (mRefreshEnable && mFirstItemIndex == 0) {
                    mHeaderView.move((int) (mCurrentY - mStartY));
                }
                break;
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                mCurrentY = event.getY();
                LogTool.logi("CListView", "UP startY = " + mStartY + ", currentY = " + mCurrentY);
                if (mRefreshEnable && mFirstItemIndex == 0 && mRecord) {
                    mHeaderView.release((int) (mCurrentY - mStartY));
                }

                mRecord = false;
                break;
        }
        return super.onTouchEvent(event);
    }

    /*
     * 刷新回调
     */
    public void refreshFinish() {
        mHeaderView.refreshFinish();
    }

    public void setOnRefreshListener(IOnRefreshListener refreshListener) {
        mRefreshEnable = (refreshListener != null);

        mHeaderView.setOnRefreshListener(refreshListener);
    }

    public interface IOnRefreshListener {
        void onRefreshStart();
    }

}

File: app/src/main/java/com/blog/demo/control/widget/CListView.kt
package com.blog.demo.control.widget

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.widget.AbsListView
import android.widget.ListView
import com.blog.demo.LogTool

class CListView(context: Context, attrs: AttributeSet?, defStyleAttr: Int) :
        ListView(context, attrs, defStyleAttr), AbsListView.OnScrollListener {

    private var mHeaderView: CListViewHeaderView
    private var mFirstItemIndex = 0

    private var mStartY = 0f
    private var mCurrentY = 0f
    private var mRefreshEnable = false
    private var mRecord = false

    constructor(context: Context) : this(context, null)

    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)

    init {
        mHeaderView = CListViewHeaderView(context)
        mHeaderView.setListView(this)
        addHeaderView(mHeaderView)
        mHeaderView.refreshFinish()

        setOnScrollListener(this)
    }

    override fun onScroll(view: AbsListView?, firstVisibleItem: Int, visibleItemCount: Int, totalItemCount: Int) {
        mFirstItemIndex = firstVisibleItem
    }

    override fun onScrollStateChanged(view: AbsListView?, scrollState: Int) {
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {    // 保证触摸的y值只记录一次
                if (!mRecord) {
                    mStartY = event.y
                    mRecord = true
                }
            }
            MotionEvent.ACTION_MOVE -> {
                if (!mRecord) {
                    mStartY = event.y
                    mRecord = true
                }
                mCurrentY = event.y
                LogTool.logi("CListView", "MOVE startY = $mStartY, currentY = $mCurrentY")
                if (mRefreshEnable && mFirstItemIndex == 0) {
                    mHeaderView.move((mCurrentY - mStartY).toInt())
                }
            }
            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                mCurrentY = event.y
                LogTool.logi("CListView", "UP startY = $mStartY, currentY = $mCurrentY")
                if (mRefreshEnable && mFirstItemIndex == 0 && mRecord) {
                    mHeaderView.release((mCurrentY - mStartY).toInt().toFloat())
                }
                mRecord = false
            }
        }
        return super.onTouchEvent(event)
    }

    /*
     * 刷新回调
     */
    fun refreshFinish() {
        mHeaderView.refreshFinish()
    }

    fun setOnRefreshListener(refreshListener: IOnRefreshListener) {
        mRefreshEnable = refreshListener != null
        mHeaderView.setOnRefreshListener(refreshListener)
    }

    interface IOnRefreshListener {
        fun onRefreshStart()
    }

}